### PR TITLE
[Fix EI-4795] MB XDeliveryCount header does not reflect correct values

### PR DESCRIPTION
**Pull request by abeykoon**
## Purpose

[Fix EI-4795] MB XDeliveryCount header does not reflect correct values

## Goals

Fixes: https://github.com/wso2/product-ei/issues/4795

## Approach

Fix the race condition on requeing messages 

## Documentation

N/A

## Training

N/A

## Certification

N/A

## Marketing

N/A

## Automation tests

Will add integration test 

## Security checks

yes

### CODE REVIEW

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/DeliverableAndesMetadata.java [hunk 1]
```diff
@@ -251,10 +251,14 @@ public void markAsDispatchedToDeliver(UUID channelID) {
         channelInformation.addChannelStatus(ChannelMessageStatus.DISPATCHED);
 
         if (!this.isBeyondLastRollbackedMessage) {
-            channelInformation.incrementDeliveryCount();
+            int i = channelInformation.incrementDeliveryCount();
+            MessageTracer.trace(getMessageID(), getDestination(), "Message within rollback id. "
+                    + "Increased count " + i);
```
Comment (sdkottegoda) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/DeliverableAndesMetadata.java]: I suppose increased count is a bit misleading. We can use the term current count as used in the else block

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/subscription/AndesSubscription.java [hunk 2]
```diff
@@ -248,12 +248,14 @@ public void onMessageAck(long messageID) throws AndesException {
...
+     * @param isMessageBeyondLastRollback true if message was rejected after last rollback
      * @throws AndesException on a message re-schedule issue
      * @return DeliverableAndesMetadata reference of rejected message
      */
-    public DeliverableAndesMetadata onMessageReject(long messageID, boolean reQueue)
-            throws AndesException {
+    public DeliverableAndesMetadata onMessageReject(long messageID, boolean reQueue,
+                                                    boolean isMessageBeyondLastRollback) throws AndesException {
         DeliverableAndesMetadata rejectedMessage = subscriberConnection.onMessageReject(messageID);
+        rejectedMessage.setIsBeyondLastRollbackedMessage(isMessageBeyondLastRollback);
```
Comment (sdkottegoda) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/subscription/AndesSubscription.java]: Should this be done before subscriberConnection.onMessageReject?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/DeliverableAndesMetadata.java (hunk 1 above)
Comment (abeykoon) [reply modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/DeliverableAndesMetadata.java]: Fixed.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/subscription/AndesSubscription.java (hunk 2 above)
Comment (abeykoon) [reply modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/subscription/AndesSubscription.java]: What on message reject does is removing the message from outbound tracker and getting the message reference at the server side. We need to call setIsBeyondLastRollbackedMessage on that message.

**Pull request merged**